### ISSUE TITLE
PSA: Psi4 and numpy v2

### ISSUE THREAD

**Issue by loriab** [psi4/psi4#3184]
tl;dr Psi4 itself needs no changes for numpy v2, but its runtime major version does need to match the compile-time major version. Current conda-package v1.9.1 with _1 needs to be used in a np v1 env. I'll rebuild a _2 soon that can be used with np v2

longer:
Numpy v2 is up on conda-forge and so may show up in an environment near you. The situation wrt psi4 is that it needs either v1 or v2 depending on what it was compiled against. So:

* the conda package -- AVOIDABLE PROBLEM, this is compiled against numpy v1, so constrain your env. I'll do a new build against v2 after qcel is ready.
* an old objdir compiled against a v1 numpy -- no problem, just don't let any new conda commands update the env to v2 numpy. use `numpy=1* -c conda-forge` in any `conda update` or `conda install` commands.
* a fresh objdir where the numpy is v2 -- no problem, psi4 compiles just fine. Might need https://github.com/MolSSI/QCElemental/pull/338
* an old objdir compiled against a v1 numpy that now sees a v2 numpy at runtime -- PROBLEM, see signature below. downgrade the numpy with `conda install numpy=1 -c conda-forge`

The symptom of an incompatible psi4+numpy is the following.

```
!----------------------------------------------------------------------------------!
!                                                                                  !
!  Following atoms are too close: [(0, np.int64(1), np.float64(0.0)), (0,          !
!     np.int64(2), np.float64(0.0)), (1, np.int64(2), np.float64(0.0))]            !
!                                                                                  !
!----------------------------------------------------------------------------------!
```

**Comment by loriab** [comment]
UPDATE: Above holds that Psi4 can be built against NumPy v1 _or_ v2, just be sure your runtime np is the same major version as your build-time numpy, or you'll get the "atoms are too close" error. _But_ I've learned a few more constraints and pitfalls since then:

* to use np v2, you must have `pybind11>=2.12` (right now the codedeps/psi4-path-advisor setup is constraining to 2.10 or 2.11 to avoid a docs issue; working on a resolution) (signature is "atoms are too close")
* to use np v2, you must have `pint>=0.24.1` (v0.24.0 works for >=py310; v0.24.1 works for >=py39) (signature is something like sumprod not available in numpy)
* to use np v2, you must have `qcelemental>=0.28.0` (fixed since last week; qcengine doesn't need changes)
* on c-f, packages have ways or declaring what they're compatible with so that an env solves into a workable state. Those aren't in place at present for the numpy+pint and numpy+pybind11 combinations, so you need to check them yourself.
* even if your build env is np v2, beware of any pip installs that can surreptitiously sneak in a np v1
